kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/KotlinSymbolProcessing.kt

```diff
@@ -8,66 +32,112 @@ import com.intellij.openapi.util.Disposer
-    val env = KotlinCoreEnvironment.createForProduction(
+    val providers: List<SymbolProcessorProvider>
+    val env: KotlinCoreEnvironment = KotlinCoreEnvironment.createForProduction(
         Disposer.newDisposable(), compilerConfiguration,
         EnvironmentConfigFiles.JVM_CONFIG_FILES
     )
+    val project = env.project as MockProject
+    val kspCoreEnvironment = KSPCoreEnvironment(project)
 
-    val application = ApplicationManager.getApplication() as MockApplication
-    configureApplicationEnvironment(application)
+    var finished = false
+    val deferredSymbols = mutableMapOf<SymbolProcessor, List<KSAnnotated>>()
+    val ktFiles = convertFilesToKtFiles(project, compilerConfiguration.kotlinSourceRoots.map { it.path })
+    val codeGenerator: CodeGeneratorImpl
+    val processors: List<SymbolProcessor>
 
-    val files = kotlinSourceRoots
-        .sortedBy { Files.isSymbolicLink(it.toPath()) } // Get non-symbolic paths first
-        .flatMap { root -> root.walk().filter { it.isFile && it.extension == "kt" }.toList() }
-        .sortedBy { Files.isSymbolicLink(it.toPath()) } // This time is for .java files
-        .distinctBy { it.canonicalPath }
-    compilerConfiguration.addKotlinSourceRoots(files.map { it.absolutePath })
+    init {
+        configureProjectEnvironment(
+            project,
+            compilerConfiguration,
+            env::createPackagePartProvider,
+            env.projectEnvironment.environment.jarFileSystem as CoreJarFileSystem
+        )
 
-    val project = env.project as MockProject
-    val ktFiles = convertFilesToKtFiles(project, files)
-    configureProjectEnvironment(
-        project,
-        compilerConfiguration,
-        env::createPackagePartProvider,
-        env.projectEnvironment.environment.jarFileSystem as CoreJarFileSystem
-    )
-    val kspCoreEnvironment = KSPCoreEnvironment(project)
+        val ksFiles = ktFiles.map { KSFileImpl(it) }
+        val anyChangesWildcard = AnyChanges(options.projectBaseDir)
+        codeGenerator = CodeGeneratorImpl(
+            options.classOutputDir,
+            options.javaOutputDir,
+            options.kotlinOutputDir,
+            options.resourceOutputDir,
+            options.projectBaseDir,
+            anyChangesWildcard,
+            ksFiles,
+            options.incremental
+        )
+        val application = ApplicationManager.getApplication() as MockApplication
+        configureApplicationEnvironment(application)
+        providers = if (testProcessor != null) {
+            listOf(testProcessor)
+        } else {
+            val processingClasspath = options.processingClasspath
+            val classLoader =
+                URLClassLoader(processingClasspath.map { it.toURI().toURL() }.toTypedArray(), javaClass.classLoader)
+
+            ServiceLoaderLite.loadImplementations(SymbolProcessorProvider::class.java, classLoader)
+        }
+
+        processors = providers.mapNotNull { provider ->
```

Comment:
It's weird to call or even load processors in this constructor. Can you move them into `execute()`?